kubernetes/resource_kubernetes_csi_driver_test.go

```diff
@@ -0,0 +1,103 @@
+package kubernetes
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+	api "k8s.io/api/storage/v1beta1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestAccKubernetesCSIDriver_basic(t *testing.T) {
+	var conf api.CSIDriver
+	resourceName := "kubernetes_csi_driver.test"
+	name := acctest.RandomWithPrefix("tf-acc-test")
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:      func() { testAccPreCheck(t) },
+		IDRefreshName: resourceName,
+		Providers:     testAccProviders,
+		CheckDestroy:  testAccCheckKubernetesCSIDriverDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccKubernetesCSIDriverBasicConfig(name, true),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckKubernetesCSIDriverExists(resourceName, &conf),
+					resource.TestCheckResourceAttr(resourceName, "metadata.0.name", name),
+					resource.TestCheckResourceAttr(resourceName, "spec.0.attach_required", "true"),
+					resource.TestCheckResourceAttr(resourceName, "spec.0.pod_info_on_mount", "true"),
+					resource.TestCheckResourceAttr(resourceName, "spec.0.volume_lifecycle_modes.0", "Ephemeral"),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func testAccCheckKubernetesCSIDriverDestroy(s *terraform.State) error {
+	conn, err := testAccProvider.Meta().(KubeClientsets).MainClientset()
+	if err != nil {
+		return err
+	}
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "kubernetes_csi_driver" {
+			continue
+		}
+		name := rs.Primary.ID
+		resp, err := conn.StorageV1beta1().CSIDrivers().Get(name, meta_v1.GetOptions{})
+		if err == nil {
+			if resp.Name == rs.Primary.ID {
+				return fmt.Errorf("CSIDriver still exists: %s", rs.Primary.ID)
+			}
+		}
+	}
+
+	return nil
+}
+
+func testAccCheckKubernetesCSIDriverExists(n string, obj *api.CSIDriver) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[n]
+		if !ok {
+			return fmt.Errorf("Not found: %s", n)
+		}
+
+		conn, err := testAccProvider.Meta().(KubeClientsets).MainClientset()
+		if err != nil {
+			return err
+		}
+
+		name := rs.Primary.ID
+		out, err := conn.StorageV1beta1().CSIDrivers().Get(name, meta_v1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		*obj = *out
+		return nil
+	}
+}
+
```

Comment:
We should also have an import test, see [config_map_test.go](https://github.com/hashicorp/terraform-provider-kubernetes/blob/master/kubernetes/resource_kubernetes_config_map_test.go#L116) for an example.

Reply:
the basic test has an import check, do you want me spit this to its own test?

Reply:
Ah so it does - yeah we should split them to make it more clear what's being tested.

Reply:
I had a look and it seems like most tests do have a `_importBasic` test case rather than doing it as a check in the basic test case - we should go with that. The reason for this is that you get a much clearer signal on what has failed when you run the tests. If you bake it into the basic test case you have to look a little bit closer to see that it is actually a problem with import.

This is particularly useful for me because behind the scenes I kick off many test builds against different Kubernetes versions and different cloud providers for these PRs so I look a lot of test output 😄

Reply:
Np. will change now.

Reply:
done:
```
--- PASS: TestAccKubernetesCSIDriver_basic (3.85s)
--- PASS: TestAccKubernetesCSIDriver_importBasic (1.89s)
```